### ISSUE TITLE
OOM with simple scan_csv + sink_parquet

### ISSUE THREAD

**Issue by raayu83** [pola-rs/polars#13458]
### Checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
    lf = pl.scan_csv(
        csv_folder.glob("**/*"),
        low_memory=True,
    )
    lf.sink_parquet("test.parquet)
```
If you can point me to any publicly available CSV files of 10GB or larger I'll write a complete example.

### Log output

```shell
UN STREAMING PIPELINE
union -> parquet_sink
RefCell { value: [] }
STREAMING CHUNK SIZE: 99996 rows
```

### Issue description

When reading a large csv with ~450 Million rows and three columns with scan_csv and writing them with sink_parquet, more than 50GB of memory are needed.
I'm running this in Kubernetes in case it makes a difference.

### Expected behavior

Memory usage should be low enough for 8GB memory to be sufficient

### Installed versions

<details>

```
--------Version info---------
Polars:               0.20.3
Index type:           UInt32
Platform:             Linux-5.15.136-flatcar-x86_64-with-glibc2.31
Python:               3.9.2 (default, Feb 28 2021, 17:03:44) 
[GCC 10.2.1 20210110]
----Optional dependencies----
adbc_driver_manager:  <not installed>
cloudpickle:          3.0.0
connectorx:           <not installed>
deltalake:            <not installed>
fsspec:               <not installed>
gevent:               <not installed>
hvplot:               <not installed>
matplotlib:           3.8.2
numpy:                1.26.3
openpyxl:             3.1.2
pandas:               2.1.4
pyarrow:              14.0.2
pydantic:             <not installed>
pyiceberg:            <not installed>
pyxlsb:               <not installed>
sqlalchemy:           2.0.25
xlsx2csv:             <not installed>
xlsxwriter:           3.1.9
```

</details>

**Comment by ritchie46** [comment]
Hard to reproduce like this. Can you tell me more what is your schema. How does memory increase overtime? Why do you need to glob? Has that influence?

**Comment by ritchie46** [comment]
Yes, maybe compile with debug symbols and get a [heaptack](https://github.com/KDE/heaptrack) report?

**Comment by dwink** [comment]
I see the same behavior. Memory seems to continually grow while streaming from csv to parquet. I’ve tried to turn off all optimizations as well as compression, and adjusted row_group_size.

I’ll see if I can profile further as well.

**Comment by dwink** [comment]
I was able to reproduce this in Rust, creating a LazyCsvReader and then calling .sink_parquet() on the resulting LazyFrame.

**Comment by edsu** [comment]
I'm seeing this memory behavior too with a simple example with a 12 GB CSV file using polars v1.2.1:

```python
lf = pl.scan_csv('test.csv')
lf.sink_parquet('test.parquet')
```

Watching RSS memory usage with `ps` while that runs shows that it peaks at 21 GB. 

Experimenting with reading with a `schema`, and writing using `compression` and `row_group_size` had no effect. It also doesn't appear related to writing the Parquet file since `sink_csv()` behaves the same.

I don't know if this is a regression, but it does feel like this behavior is not described correctly in the documentation.

**Comment by coastalwhite** [comment]
Streaming engine is completely redone. Closing.
